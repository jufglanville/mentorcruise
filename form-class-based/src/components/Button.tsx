import React, { Component, ReactNode } from 'react';
import styled from 'styled-components';

interface Props {
  children: ReactNode;
  onClick: () => void;
}

interface State {
  clicked: boolean;
}

class Button extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { clicked: false };
  }
  render() {
    const { children, onClick } = this.props;
    return <Btn onClick={onClick}>{children}</Btn>;
  }
}

const Btn = styled.button`
  background-color: var(--green);
  color: var(--white);
  padding: 1rem;
  border-radius: 0.3rem;
  border: none;
  font-size: 0.8rem;
  font-weight: 500;
  text-transform: uppercase;
  text-align: center;
  cursor: pointer;
  transition: all 0.2s ease-in-out;
  box-shadow: 0 0.5rem 1rem rgba(0, 0, 0, 0.1);
  letter-spacing: 0.05rem;
  &:hover {
    opacity: 0.8;
  }
`;

export default Button;
