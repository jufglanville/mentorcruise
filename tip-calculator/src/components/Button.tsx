import React from 'react';
import styled from 'styled-components';

type TranscientProps = {
  $selected: boolean;
  disabled: boolean;
};

interface Props {
  children: React.ReactNode;
  disabled?: boolean;
  selected?: boolean;
  onClick: () => void;
}

const Button = ({
  children,
  onClick,
  selected = false,
  disabled = false,
}: Props) => {
  return (
    <ButtonElement $selected={selected} disabled={disabled} onClick={onClick}>
      {children}
    </ButtonElement>
  );
};

const ButtonElement = styled.button<TranscientProps>`
  background-color: ${({ $selected }) =>
    $selected ? 'var(--primary)' : 'var(--dark-cyan)'};
  color: ${({ $selected }) =>
    $selected ? 'var(--dark-cyan)' : 'var(--white)'};
  border: none;
  border-radius: 4px;
  padding: 8px 16px;
  font-size: 24px;
  font-weight: 700;
  text-transform: uppercase;
  text-align: center;
  letter-spacing: 0.5px;
  cursor: pointer;
  &:hover {
    color: var(--dark-cyan);
    background-color: var(--light-grayish-cyan);
  }
  &:disabled {
    background-color: var(--primary);
    color: var(--dark-cyan);
    opacity: 0.5;
    cursor: not-allowed;
  }
`;

export default Button;
