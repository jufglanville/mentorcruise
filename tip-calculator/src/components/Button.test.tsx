import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import Button from './Button';
import React from 'react';

const mockOnClick = jest.fn();

describe('Button component', () => {
  it('Renders a Button with the correct text', () => {
    render(
      <Button selected={true} onClick={mockOnClick}>
        Click Me
      </Button>
    );
    const button = screen.getByText(/click me/i);
    expect(button).toBeInTheDocument();
  });

  it('Allows us to click the button', () => {
    render(
      <Button selected={true} onClick={mockOnClick}>
        Click Me
      </Button>
    );
    const button = screen.getByText(/click me/i);
    fireEvent.click(button);
    expect(mockOnClick).toHaveBeenCalled();
  });

  it('Prevents button being clicked if disabled', () => {
    render(
      <Button selected={true} onClick={mockOnClick} disabled={true}>
        Click Me
      </Button>
    );
    const button = screen.getByText(/click me/i);
    expect(button).toBeDisabled();
  });

  it('Applies the correct styles when selected', () => {
    const {container} = render(
      <Button selected={true} onClick={mockOnClick}>
        Click Me
      </Button>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Applies the correct default styles', () => {
    const {container} = render(
      <Button selected={false} onClick={mockOnClick}>
        Click Me
      </Button>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Applies the correct styles when disabled', () => {
    const {container} = render(
      <Button selected={false} onClick={mockOnClick} disabled={true}>
        Click Me
      </Button>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
