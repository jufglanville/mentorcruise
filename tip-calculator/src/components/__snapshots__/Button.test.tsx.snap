// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Applies the correct default styles 1`] = `
<button
  class="sc-aXZVg egkSUg"
>
  Click Me
</button>
`;

exports[`Button component Applies the correct styles when disabled 1`] = `
<button
  class="sc-aXZVg egkSUg"
  disabled=""
>
  Click Me
</button>
`;

exports[`Button component Applies the correct styles when selected 1`] = `
<button
  class="sc-aXZVg hdPvyL"
>
  Click Me
</button>
`;
